File: api/schemas/reactors.py
from pydantic import BaseModel, Field


class BatchInfo(BaseModel):
    hours_in_reactor: float | None = None
    qc_result: str
    qc_open: bool
    hours_at_qc: float | None = None


class ReactorCard(BaseModel):
    id: int
    reactor_name: str
    capacity_l: float
    asset_tag: str
    bay_marker: str
    current_resin: str | None
    assigned_pump: str | None
    is_idle: bool
    remaining_l: float
    remaining_kg: float
    fill_pct: float
    lot: str
    svg: str
    batch: BatchInfo | None = None
    can_mark_empty: bool


class VesselTypeOption(BaseModel):
    value: str
    label: str
    help: str


class ManageOptionsOut(BaseModel):
    vessel_types: list[VesselTypeOption]
    resins: list[str]
    pumps: list[str]
    bulk_enabled: bool


class ManageReactorRow(BaseModel):
    id: int
    reactor_name: str
    capacity_l: float
    vessel_type: str
    asset_tag: str
    bay_marker: str
    assigned_pump: str
    current_resin: str


class AddReactorRequest(BaseModel):
    reactor_name: str
    max_capacity_l: float = Field(gt=0)
    vessel_type: str = ""
    asset_tag: str = ""
    bay_marker: str = ""


class UpdateReactorRequest(BaseModel):
    vessel_type: str
    asset_tag: str = ""
    bay_marker: str = ""
    assigned_pump: str = ""   # "" means "not set"
    current_resin: str = ""  # "" means "not set"


class BulkCandidate(BaseModel):
    id: int
    reactor_name: str
    label: str
    current_resin: str


class DrawInfo(BaseModel):
    capacity_l: float
    current_resin: str
    lot: str
    remaining_l: float
    density_kg_l: float


class BulkPourRequest(BaseModel):
    containers: int = Field(ge=1, le=99)
    amount_each: float = Field(ge=0)
    unit: str = "L"
    note: str = ""


class MarkEmptyOut(BaseModel):
    ok: bool

File: frontend/src/api/reactors.ts
import { api } from './client'

export interface BatchInfo {
  hours_in_reactor: number | null
  qc_result: string
  qc_open: boolean
  hours_at_qc: number | null
}

export interface ReactorCard {
  id: number
  reactor_name: string
  capacity_l: number
  asset_tag: string
  bay_marker: string
  current_resin: string | null
  assigned_pump: string | null
  is_idle: boolean
  remaining_l: number
  remaining_kg: number
  fill_pct: number
  lot: string
  svg: string
  batch: BatchInfo | null
  can_mark_empty: boolean
  can_manage: boolean
}

export interface VesselTypeOption {
  value: string
  label: string
  help: string
}

export interface ManageOptions {
  vessel_types: VesselTypeOption[]
  resins: string[]
  pumps: string[]
  bulk_enabled: boolean
}

export interface ManageReactorRow {
  id: number
  reactor_name: string
  capacity_l: number
  vessel_type: string
  asset_tag: string
  bay_marker: string
  assigned_pump: string
  current_resin: string
}

export interface BulkCandidate {
  id: number
  reactor_name: string
  label: string
  current_resin: string
}

export interface DrawInfo {
  capacity_l: number
  current_resin: string
  lot: string
  remaining_l: number
  density_kg_l: number
}

export const reactorsApi = {
  fleet: () => api.get<ReactorCard[]>('/reactors/fleet'),
  manageOptions: () => api.get<ManageOptions>('/reactors/manage-options'),
  manageList: () => api.get<ManageReactorRow[]>('/reactors/manage-list'),
  add: (body: { reactor_name: string; max_capacity_l: number; vessel_type: string; asset_tag: string; bay_marker: string }) =>
    api.post<{ ok: boolean }>('/reactors', body),
  remove: (id: number) => api.del<{ ok: boolean }>(`/reactors/${id}`),
  update: (id: number, body: { vessel_type: string; asset_tag: string; bay_marker: string; assigned_pump: string; current_resin: string }) =>
    api.put<{ ok: boolean }>(`/reactors/${id}`, body),
  bulkCandidates: () => api.get<BulkCandidate[]>('/reactors/bulk-candidates'),
  drawInfo: (id: number) => api.get<DrawInfo>(`/reactors/${id}/draw-info`),
  bulkPour: (id: number, body: { containers: number; amount_each: number; unit: string; note: string }) =>
    api.post<{ ok: boolean; litres: number }>(`/reactors/${id}/bulk-pour`, body),
  markEmpty: (id: number) => api.post<{ ok: boolean }>(`/reactors/${id}/mark-empty`),
  markFilled: (id: number, body: { resin_type: string; lot_number: string; filled_at: string | null; note: string }) =>
    api.post<{ ok: boolean }>(`/reactors/${id}/mark-filled`, body),
  reconcile: (id: number, body: { mode: 'percent' | 'liters'; value: number; notes: string }) =>
    api.post<{ ok: boolean }>(`/reactors/${id}/reconcile`, body),
}
